File: 9-12.cpp
#include<bits/stdc++.h>
using namespace std;

vector<int> E[20];
int N, answer;

void solve(int state, int sheep, int wolf, vector<int>& info) {
    if (sheep > answer)
        answer = sheep;

    for (int u = 0; u < N; u++) {
        if ((state & (1 << u)) == 0) continue;

        for (int i = 0; i < E[u].size(); i++) {
            int v = E[u][i];
            if (state & (1 << v)) continue;

            if (info[v] == 0) {
                solve(state | (1 << v), sheep + 1, wolf, info);
            }
            else {
                if (sheep > wolf + 1) {
                    solve(state | (1 << v), sheep, wolf + 1, info);
                }
            }
        }
    }
}

int solution(vector<int> info, vector<vector<int>> edges) {
    N = info.size();
    for (int i = 0; i < edges.size(); i++) {
        int p = edges[i][0], c = edges[i][1];
        E[p].push_back(c);
    }

    solve(1, 1, 0, info);
    return answer;
}

File: 9-12.py
def solve(state, sheep, wolf, info, E):
    ret = sheep

    for u in range(len(info)):
        if (state & (1 << u)) == 0:
            continue

        for v in E[u]:
            if state & (1 << v):
               continue

            if info[v] == 0:
                ret = max(ret, solve(state | (1 << v), sheep + 1, wolf, info, E))
            else:
                if sheep > wolf + 1:
                    ret = max(ret, solve(state | (1 << v), sheep, wolf + 1, info, E))
    return ret

def solution(info, edges):
    E = [[] for _ in range(len(info))]
    for p, c in edges:
        E[p].append(c)

    return solve(1, 1, 0, info, E)